### PR TITLE
feat: Feature/list tables page size

### LINKED ISSUE #173: dataset.list_tables max_results parameter
I'm using great_expectations to test my data in BigQuery and the library introspects the entire database/schema for tables, meaning they use the function `sa.inspect(engine).get_table_names()`, because I have alot of tables this process is very very slow, I want to change the `max_results` parameter passed on to [dataset.list_tables operation](https://cloud.google.com/bigquery/docs/reference/rest/v2/tables/list), when tested it increased performance dramatically

### PR DESCRIPTION
**Pull request by OmriBromberg**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [V] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-sqlalchemy/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [V] Ensure the tests and linter pass
- [V] Code coverage does not decrease (if any source code was changed)
- [V] Appropriate docs were updated (if necessary)

Fixes #173 🦕

Hey guys, I explained in the issue a bit but I wanted to implement it so users would be able to change max_results @ [dataset.list_tables](https://cloud.google.com/bigquery/docs/reference/rest/v2/tables/list) like the arraysize parameter so I copied how it is treated in the parse_url area and updated the tests and docs for it, the new parameter for it is called `list_tables_page_size`

also I saw that on [line 606](https://github.com/googleapis/python-bigquery-sqlalchemy/blob/f5adfc091f37d531921374768742a199a8c41420/pybigquery/sqlalchemy_bigquery.py#L606)
there is this expression `self.arraysize = self.arraysize or arraysize` which always takes the default arraysize instead of the parsed querystring arraysize because self.arraysize is always initialized to a default value, so I also reversed it to be as so `self.arraysize = arraysize or self.arraysize`, now the querystring arraysize will take precedence over the default value

### CODE REVIEW

**Comment by OmriBromberg** [comment]
@googlebot I signed it!

**Review by jimfulton — COMMENTED** [review]
This is a great start. In addition to adding a feature, you found two bugs!

This needs test of the logic that fetches lists of tables.  A unit test would be fine and should be straightforward to write.

File: pybigquery/sqlalchemy_bigquery.py [hunk 1]
```diff
@@ -601,9 +603,11 @@ def create_connect_args(self, url):
             arraysize,
             credentials_path,
             default_query_job_config,
+            list_tables_page_size,
         ) = parse_url(url)
 
-        self.arraysize = self.arraysize or arraysize
+        self.arraysize = arraysize or self.arraysize
```
[the remark this answers is not in the document]
Comment (OmriBromberg) [reply pybigquery/sqlalchemy_bigquery.py:609]: sure :) ima add a test for it

File: pybigquery/sqlalchemy_bigquery.py [hunk 2]
```diff
@@ -644,7 +648,9 @@ def _get_table_or_view_names(self, connection, table_type, schema=None):
                 continue
 
             try:
-                tables = client.list_tables(dataset.reference)
+                tables = client.list_tables(
+                    dataset.reference, max_results=self.list_tables_page_size
+                )
```
[the remark this answers is not in the document]
Comment (OmriBromberg) [reply pybigquery/sqlalchemy_bigquery.py:653]: ohh, I thought the pagination was inside the list_tables method and it just blocks till it finishes, can you explain a bit more then?

**Comment by OmriBromberg** [comment]
hey, regarding [this](https://github.com/googleapis/python-bigquery-sqlalchemy/blob/f5adfc091f37d531921374768742a199a8c41420/pybigquery/sqlalchemy_bigquery.py#L631-L636) can you explain this logic to me? why shouldn't it always take the "table_id" property? gives us inconsistencies

File: pybigquery/sqlalchemy_bigquery.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jimfulton) [reply pybigquery/sqlalchemy_bigquery.py:653]: By default, only one page is returned.  You need to pass `page_token` to get later pages.

Here's an example:
```
>>> i = client.list_tables("riversnake", 2)
>>> [t.table_id for t in i]
['dates', 'fix667']
>>> i = client.list_tables("riversnake", 2, i.next_page_token)
>>> [t.table_id for t in i]
['foo', 'fruits']
>>> i = client.list_tables("riversnake", 2, i.next_page_token)
>>> [t.table_id for t in i]
['t0', 't1']
```

**Comment by jimfulton** [comment]
> hey, regarding [this](https://github.com/googleapis/python-bigquery-sqlalchemy/blob/f5adfc091f37d531921374768742a199a8c41420/pybigquery/sqlalchemy_bigquery.py#L631-L636) can you explain this logic to me? why shouldn't it always take the "full_table_id" property? gives us inconsistencies

That's a good question.

If you look at: https://github.com/sqlalchemy/sqlalchemy/blob/b20b6f8fe7ea0198f819a0fd68ca076b6c760054/lib/sqlalchemy/testing/suite/test_reflection.py#L537-L579

SQLAlchemy expects the results to be unqualified, which we satisfy if we have a schema.

Looking at other dialects:

MS SQL requires a schema: https://github.com/sqlalchemy/sqlalchemy/blob/b20b6f8fe7ea0198f819a0fd68ca076b6c760054/lib/sqlalchemy/dialects/mssql/base.py#L2877

Oracle assumes there's a default: https://github.com/sqlalchemy/sqlalchemy/blob/b20b6f8fe7ea0198f819a0fd68ca076b6c760054/lib/sqlalchemy/dialects/oracle/base.py#L1759-L1760

as does Postgres: https://github.com/sqlalchemy/sqlalchemy/blob/b20b6f8fe7ea0198f819a0fd68ca076b6c760054/lib/sqlalchemy/dialects/postgresql/base.py#L3488-L3490

As does sqlite, sort of: https://github.com/sqlalchemy/sqlalchemy/blob/b20b6f8fe7ea0198f819a0fd68ca076b6c760054/lib/sqlalchemy/dialects/sqlite/base.py#L1977-L1981

and so on.

So BQ is a little different because there might not be a default.

Amongst the options:

1. Error if no schema is provided and there's no default.
2. Qualify the results if we don't have a schema.  I prefer using just the data set id in this case.  `full_table_id` adds clutter IMO.  

----

Playing with this some more, you can list schemas in other projects by specifying `project_id.schema_id` (e.g.  "bigquery-public-data.census_bureau_acs").  That makes the use of a colon in `full_table_id` a little weird.

File: pybigquery/sqlalchemy_bigquery.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jimfulton) [reply pybigquery/sqlalchemy_bigquery.py:653]: I should have noted that `next_page_token` will be `None` when you've reached the end.

[the remark this answers is not in the document]
Comment (OmriBromberg) [reply pybigquery/sqlalchemy_bigquery.py:653]: ohhh I see, I'll make it work then :), also I'll try and add a test for it

[the remark this answers is not in the document]
Comment (jimfulton) [reply pybigquery/sqlalchemy_bigquery.py:653]: Thanks!

[the remark this answers is not in the document]
Comment (OmriBromberg) [reply pybigquery/sqlalchemy_bigquery.py:653]: Update:
I did some checking and found [this](https://github.com/googleapis/python-api-core/blob/c4ff747db17cdeedb0eb2741b308c7b2d2a596fb/google/api_core/page_iterator.py#L210-L249) basically the HTTPIterator has already implemented the logic for paginating over the results and outputting it in one nice iterator
The problem is in this case the parameter max_results is used once as a "page_size" [here](https://github.com/googleapis/python-api-core/blob/c4ff747db17cdeedb0eb2741b308c7b2d2a596fb/google/api_core/page_iterator.py#L402-L426) and once as a "max_results for the iterator" [here](https://github.com/googleapis/python-api-core/blob/c4ff747db17cdeedb0eb2741b308c7b2d2a596fb/google/api_core/page_iterator.py#L387-L389)
I don't know if this behavior is intended or not, but when you pass max_results 2147483647 (max int32) for example, the API doesn't actually reject our API call but just uses the max it allows (although undocumented after further digging found it to be 1000), so maybe we can rename our parameter or change code over at HTTPIterator or just do the pagination ourselves although it seems a bit odd, IDK whats the best practice here, please share some ideas so I'll know what's the best solution here
Also thanks for the quick replies :)

[the remark this answers is not in the document]
Comment (jimfulton) [reply pybigquery/sqlalchemy_bigquery.py:653]: So much for the quick reply ... ;)  Thanks for your patience.

This logic is very confusing and the documentation doesn't help. I suspect this was partially refactored at some point in the past.

I think this API is broken.  There's no correct way to get all of the data and a batch size of 1000 without the client doing pagination, and I don't think that's the current intent.

I'm new to this project, and I don't know the history.  I'm going to need to ask around a bit.

Stay tuned...

[the remark this answers is not in the document]
Comment (jimfulton) [reply pybigquery/sqlalchemy_bigquery.py:653]: I created an issue: https://github.com/googleapis/python-api-core/issues/194.

Let's see what happens there.

[the remark this answers is not in the document]
Comment (OmriBromberg) [reply pybigquery/sqlalchemy_bigquery.py:653]: OK, sounds awesome, I can chip in and help in the other PR if there is any, just let me know :)

**Comment by jimfulton** [comment]
The bigquery client now accepts a `page_size` parameter, so this PR wants to use that.

When your;re ready for me to review, please merge master to get rid of many spurious changes. :)

**Review by jimfulton — APPROVED** [review]

**Comment by jimfulton** [comment]
Thanks!

**Pull request merged**